packages/codegen.go/src/time.ts

```diff
@@ -51,51 +51,77 @@ export async function generateTimeHelpers(codeModel: go.CodeModel, packageName?:
+          if (go.isBodyParameter(param) || unwrappedParam.dateTimeFormat === 'timeRFC3339' || unwrappedParam.dateTimeFormat === 'timeUnix') {
+            setHelper(unwrappedParam.dateTimeFormat);
+          }
+        }
+      }
+    }
 
-  for (const client of codeModel.clients) {
-    for (const method of client.methods) {
-      for (const param of method.parameters) {
-        const unwrappedParam = recursiveUnwrapMapSlice(param.type);
-        if (!go.isTimeType(unwrappedParam)) {
+    for (const model of codeModel.models) {
+      for (const field of values(model.fields)) {
+        const unwrappedField = recursiveUnwrapMapSlice(field.type);
+        if (!go.isTimeType(unwrappedField)) {
           continue;
         }
-        // for body params, the helpers are always required.
-        // for header/path/query params, the conversion happens in place. the only
-        // exceptions are for timeRFC3339 and timeUnix
-        // TODO: clean this up when moving to DateTime type in azcore
-        if (go.isBodyParameter(param) || (unwrappedParam.dateTimeFormat === 'timeRFC3339' || unwrappedParam.dateTimeFormat === 'timeUnix')) {
-          setHelper(unwrappedParam.dateTimeFormat);
+        if (getSerDeFormat(model, codeModel) === 'JSON') {
+          // needsSerDeHelpers helpers are for JSON only
+          needsSerDeHelpers = true;
         }
+        setHelper(unwrappedField.dateTimeFormat);
       }
     }
-  }
 
-  for (const respEnv of codeModel.responseEnvelopes) {
-    if (!respEnv.result || !go.isMonomorphicResult(respEnv.result) || respEnv.result.format !== 'JSON') {
-      continue;
+    for (const respEnv of codeModel.responseEnvelopes) {
+      if (!respEnv.result || !go.isMonomorphicResult(respEnv.result) || respEnv.result.format !== 'JSON') {
+        continue;
+      }
+      const unwrappedResult = recursiveUnwrapMapSlice(respEnv.result.monomorphicType);
+      if (!go.isTimeType(unwrappedResult)) {
+        continue;
+      }
+      setHelper(unwrappedResult.dateTimeFormat);
     }
-    const unwrappedResult = recursiveUnwrapMapSlice(respEnv.result.monomorphicType);
-    if (!go.isTimeType(unwrappedResult)) {
-      continue;
+  } else {
+    for (const client of codeModel.clients) {
+      for (const method of client.methods) {
+        for (const param of method.parameters) {
+          if (go.isTimeType(param.type) && go.isBodyParameter(param)) {
+            setHelper(param.type.dateTimeFormat);
+          }
+        }
+      }
     }
-    setHelper(unwrappedResult.dateTimeFormat);
-  }
 
-  // needsSerDeHelpers is only required when time.Time is a struct field
-  let needsSerDeHelpers = false;
-
-  for (const model of codeModel.models) {
-    for (const field of values(model.fields)) {
-      const unwrappedField = recursiveUnwrapMapSlice(field.type);
-      if (!go.isTimeType(unwrappedField)) {
-        continue;
+    for (const respEnv of codeModel.responseEnvelopes) {
+      for (const header of respEnv.headers) {
```

Comment:
If checking response headers is necessary for fakes, why isn't it necessary for non-fakes?

Reply:
For fakes, it uses the [`helpers.formatValue`](https://github.com/Azure/autorest.go/blob/4cc6519fd423c84bfff8f0f0c3d53bb5b88f0a58/packages/codegen.go/src/fake/servers.ts#L399C85-L399C134) function to help generate code for the response headers. That function adds the time helpers for `timeUnix` and `timeRFC3339`.

For non-fakes, it uses the [`formatHeaderResponseValue` ](https://github.com/Azure/autorest.go/blob/4cc6519fd423c84bfff8f0f0c3d53bb5b88f0a58/packages/codegen.go/src/operations.ts#L289) function to generate code for response headers. That functions does not use any time helpers.

Reply:
Ugh, seems like we should consolidate this at some point. Thanks for the explanation.